File: apps/backend/src/wallet/wallet.controller.ts
import { Body, Controller, Get, Post, Req, UseGuards } from '@nestjs/common';
import { IsString } from 'class-validator';
import { WalletService } from './wallet.service';
import { JwtAuthGuard } from '../auth/jwt-auth.guard';
import type { RequestUser } from '../auth/jwt.strategy';

class TopupDto {
  @IsString()
  packageId: string;
}

@Controller('wallet')
@UseGuards(JwtAuthGuard)
export class WalletController {
  constructor(private walletService: WalletService) {}

  @Get()
  getWallet(@Req() req: { user: RequestUser }) {
    return this.walletService.getWallet(req.user.userId);
  }

  @Post('topup')
  topup(@Body() dto: TopupDto, @Req() req: { user: RequestUser }) {
    return this.walletService.topup(req.user.userId, dto.packageId);
  }
}

File: apps/backend-java/src/main/java/com/pawpals/backend/wallet/WalletController.java
package com.pawpals.backend.wallet;

import com.pawpals.backend.auth.RequestUser;
import com.pawpals.backend.wallet.dto.TopupRequest;
import com.pawpals.backend.wallet.dto.WalletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/wallet")
@RequiredArgsConstructor
public class WalletController {

  private final WalletService walletService;

  @GetMapping
  public WalletResponse getWallet(@AuthenticationPrincipal RequestUser user) {
    return walletService.getWallet(user.userId());
  }

  @PostMapping("/topup")
  public WalletResponse topup(@Valid @RequestBody TopupRequest dto, @AuthenticationPrincipal RequestUser user) {
    return walletService.topup(user.userId(), dto.packageId());
  }
}
